src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -0,0 +1,418 @@
+        let batches = (events.Length - 1)/ (int batchSize) + 1
+        let action = "Load"
+        let evt = Log.Metric.Batch (batches, reqMetric)
+        (log |> Log.prop "bytes" bytes |> Log.event evt).Information(
+            "Mdb{action:l} stream={stream} count={count}/{batches} version={version}",
+            action, streamName, count, batches, version)
+
+    let logLastEventRead streamName t events version (log: ILogger) =
+        let bytes = resolvedEventBytes events
+        let count = events.Length
+        let reqMetric : Log.Measurement = { stream = streamName; interval = t; bytes = bytes; count = count}
+        let evt = Log.Metric.ReadLast reqMetric
+        (log |> Log.prop "bytes" bytes |> Log.event evt).Information(
+            "Mdb{action:l} stream={stream} count={count} version={version}",
+            "ReadL", streamName, count, version)
+
+    let loadLastEvent (log : ILogger) retryPolicy (reader : MessageDbReader) requiresLeader streamName
+        : Async<int64 * ITimelineEvent<EventBody> array> = async {
+        let! ct = Async.CancellationToken
+        let read _ = readLastEventAsync reader streamName requiresLeader ct |> Async.AwaitTaskCorrect
+
+        let! t, page = Log.withLoggedRetries retryPolicy "readAttempt" read log |> Stopwatch.Time
+
+        log |> logLastEventRead streamName t page.Messages page.LastVersion
+        return page.LastVersion, page.Messages }
+
+    let loadForwardsFrom (log : ILogger) retryPolicy reader batchSize maxPermittedBatchReads streamName startPosition requiresLeader
+        : Async<int64 * ITimelineEvent<EventBody> array> = async {
+        let mergeBatches (batches : AsyncSeq<int64 * ITimelineEvent<EventBody> array>) = async {
+            let mutable versionFromStream = -1L
+            let! (events : ITimelineEvent<EventBody> array) =
+                batches
+                |> AsyncSeq.map (fun (reportedVersion, events)-> versionFromStream <- max reportedVersion versionFromStream; events)
+                |> AsyncSeq.concatSeq
+                |> AsyncSeq.toArrayAsync
+            let version = versionFromStream
+            return version, events }
+        let call pos = loggedReadSlice reader streamName batchSize pos requiresLeader
+        let retryingLoggingReadSlice pos = Log.withLoggedRetries retryPolicy "readAttempt" (call pos)
+        let log = log |> Log.prop "batchSize" batchSize |> Log.prop "stream" streamName
+        let batches : AsyncSeq<int64 * ITimelineEvent<EventBody> array> = readBatches log retryingLoggingReadSlice maxPermittedBatchReads startPosition
+        let! t, (version, events) = mergeBatches batches |> Stopwatch.Time
+        log |> logBatchRead streamName t events batchSize version
+        return version, events }
+
+module private Token =
+    let create streamVersion : StreamToken =
+        { value = box streamVersion
+          version = streamVersion
```

Comment:
This is an oversight on my part... This needs to be streamVersion+1 due to https://github.com/jet/equinox/blob/4b0790981834d6a31056e995f2e6cd5b584878a3/src/Equinox/Decider.fs#L269-L273

This mess is due to the historic 'arguable' misdesign of having a stream with one event be 'Version' 0 in ESDB (that and the empty vs not created vs no events magic values). 

(Cosmos and Dynamo use 0 for an empty stream - i.e. the version is `events.Length`)

If MessageDb is -1 based, then the chances are the only change required is adding 1 to the internal value (and adding the warning). If it's not, then we should take this opportunity to use 0-based indexes to make it all less mindbending.

https://github.com/jet/equinox/blob/master/src/Equinox.SqlStreamStore/SqlStreamStore.fs#L298

The other obvious thing here is that the tests should have caught this...
using Decider.QueryEx, the Version can be obtained by supplying a `mapResult`